daemon/cmd/ipcache.go

```diff
@@ -87,3 +105,144 @@ func containsSubnet(outer, inner net.IPNet) bool {
+	nidsToReserve := make([]identity.NumericIdentity, 0, len(d.restoredCIDRs))
+
+	for prefix, nid := range d.restoredCIDRs {
+		nidsToReserve = append(nidsToReserve, nid)
+		metaUpdates = append(metaUpdates, ipcache.MU{
+			Prefix:   prefix,
+			Source:   source.Restored,
+			Resource: restoredCIDRResource,
+			Metadata: []ipcache.IPMetadata{ipcachetypes.RequestedIdentity(nid)},
+		})
+	}
+	for _, prefix := range ingressIPs {
+		metaUpdates = append(metaUpdates, ipcache.MU{
+			Prefix:   prefix,
+			Source:   source.Restored,
+			Resource: ingressResource,
+			Metadata: []ipcache.IPMetadata{labels.LabelIngress},
+		})
+
+		// Set any restored ingress IPs back on the LocalNode object
+		d.nodeLocalStore.Update(func(n *node.LocalNode) {
+			addr := prefix.Addr()
+			if addr.Is4() {
+				n.IPv4IngressIP = addr.AsSlice()
+			} else {
+				n.IPv6IngressIP = addr.AsSlice()
+			}
+		})
+	}
+	if len(ingressIPs) > 0 {
+		log.WithField(logfields.Ingress, ingressIPs).Info("Restored ingress IPs")
+	}
+
+	// Insert the batched changes in to the ipcache.
+	// Even though the ipcache map hasn't been initialized yet, this is
+	// safe to do so, because the ipcache's apply controller is currently
+	// paused.
+	d.ipcache.IdentityAllocator.ReserveLocalIdentities(nidsToReserve)
+	d.ipcache.UpsertMetadataBatch(metaUpdates...)
+
+	return nil
+}
+
+// releaseRestoredCIDRS removes the placeholder metadata that was inserted
+// in to the ipcache when local identities were restored.
+// Any identities actually in use will still exist after this.
+//
+// This should be called after a grace period (default 10 minutes, set
+// by --identity-restore-grace-period).
+// This grace period is needed when running on an external workload
+// where policy synchronization is not done via k8s. Also in k8s
+// case it is prudent to allow concurrent endpoint regenerations to
+// (re-)allocate the restored identities before we release them.
+//
+// Any CIDRs still in use after the grace period will have other sources
+// of metadata in the ipcache, and thus will remain. CIDRs for which
+// restoration was the only source of metadata will be deallocated.
+func (d *Daemon) releaseRestoredCIDRs() {
+	defer func() {
+		// release the memory held by restored CIDRs
+		d.restoredCIDRs = nil
+	}()
+	if len(d.restoredCIDRs) == 0 {
+		return
+	}
+
+	log.WithField(logfields.Count, len(d.restoredCIDRs)).Info("Removing identity reservations for restored CIDR identities")
+	updates := make([]ipcache.MU, 0, len(d.restoredCIDRs))
+	nids := make([]identity.NumericIdentity, 0, len(d.restoredCIDRs))
+	for prefix, nid := range d.restoredCIDRs {
+		nids = append(nids, nid)
+		updates = append(updates, ipcache.MU{
+			Prefix:   prefix,
+			Resource: restoredCIDRResource,
+			Metadata: []ipcache.IPMetadata{ipcachetypes.RequestedIdentity(0)},
+		})
+	}
+
+	d.ipcache.RemoveMetadataBatch(updates...)
+	d.ipcache.UnreserveLocalIdentities(nids)
```

Comment:
The corresponding 

```
d.ipcache.IdentityAllocator.ReserveLocalIdentities(nidsToReserve)
```

is called on the `IdentityAllocator`, in the highlighted line it is acting directly on the ipcache. Is that asymmetry intended?

Reply:
No, definitely not intended. How did that ever work!? (there are some strange interfaces afoot that I don't entirely understand).